gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.data.management.dataset.DatasetUtils;
+import gobblin.data.management.util.PathUtils;
+import gobblin.util.FileListUtils;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+
+import org.apache.hadoop.fs.FileChecksum;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
+
+
+/**
+ * Implementation of {@link CopyableDataset} that creates a {@link CopyableFile} for every file that is a descendant if
+ * the root directory.
+ */
+public class RecursiveCopyableDataset extends SinglePartitionCopyableDataset {
+
+  public static final String TARGET_DIRECTORY = "gobblin.copyable.dataset.target.directory";
+
+  private final Path rootPath;
+  private final FileSystem _fs;
+  private final Path targetDirectory;
+  private final Properties properties;
+  private LoadingCache<Path, OwnerAndPermission> ownerAndPermissionCache;
+  private final PathFilter pathFilter;
+
+  public RecursiveCopyableDataset(FileSystem fs, Path rootPath, Properties properties) {
+
+    Preconditions.checkArgument(
+        !Strings.isNullOrEmpty(properties.getProperty(ConfigurationKeys.DATA_PUBLISHER_FINAL_DIR)), "Missing property "
```

Comment:
In case we want to support multiple datasets with different publish final directories we can pass the publisher final directory through the constructor instead of a properties object